File: leetcode/0987-vertical-order-traversal-of-a-binary-tree/solution.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        triples = []

        def dfs(node: Optional[TreeNode], row: int, col: int) -> None:
            if node is None:
                return
            triples.append((col, row, node.val))
            dfs(node.left, row + 1, col - 1)
            dfs(node.right, row + 1, col + 1)

        dfs(root, 0, 0)
        triples.sort()

        result = []
        current_col = None
        for col, row, val in triples:
            if col != current_col:
                result.append([])
                current_col = col
            result[-1].append(val)

        return result

File: leetcode/0987-vertical-order-traversal-of-a-binary-tree/solution.cpp
#include <vector>
#include <algorithm>
#include <tuple>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        vector<tuple<int, int, int>> triples; // (col, row, val)
        dfs(root, 0, 0, triples);

        sort(triples.begin(), triples.end());

        vector<vector<int>> result;
        bool haveCol = false;
        int currentCol = 0;
        for (auto& [col, row, val] : triples) {
            if (!haveCol || col != currentCol) {
                result.push_back({});
                currentCol = col;
                haveCol = true;
            }
            result.back().push_back(val);
        }

        return result;
    }

private:
    void dfs(TreeNode* node, int row, int col, vector<tuple<int, int, int>>& triples) {
        if (node == nullptr) {
            return;
        }
        triples.push_back({col, row, node->val});
        dfs(node->left, row + 1, col - 1, triples);
        dfs(node->right, row + 1, col + 1, triples);
    }
};
